pkg/operation/shoot/shoot.go

```diff
@@ -41,82 +40,156 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// New takes a <k8sGardenClient>, the <k8sGardenCoreInformers> and a <shoot> manifest, and creates a new Shoot representation.
-// It will add the CloudProfile, the cloud provider secret, compute the internal cluster domain and identify the cloud provider.
-func New(k8sGardenClient kubernetes.Interface, k8sGardenCoreInformers gardencoreinformers.Interface, shoot *gardencorev1beta1.Shoot, projectName string, disableDNS bool, internalDomain *garden.Domain, defaultDomains []*garden.Domain) (*Shoot, error) {
-	var (
-		secret *corev1.Secret
-		err    error
-	)
+// NewBuilder returns a new Builder.
+func NewBuilder() *Builder {
+	return &Builder{
+		shootObjectFunc: func() (*gardencorev1beta1.Shoot, error) {
+			return nil, fmt.Errorf("shoot object is required but not set")
+		},
+		cloudProfileFunc: func(string) (*gardencorev1beta1.CloudProfile, error) {
+			return nil, fmt.Errorf("cloudprofile object is required but not set")
+		},
+		shootSecretFunc: func(context.Context, client.Client, string, string) (*corev1.Secret, error) {
+			return nil, fmt.Errorf("shoot secret object is required but not set")
+		},
+	}
+}
 
-	cloudProfile, err := k8sGardenCoreInformers.CloudProfiles().Lister().Get(shoot.Spec.CloudProfileName)
-	if err != nil {
-		return nil, err
+// WithShootObject sets the shootObject attribute at the Builder.
+func (b *Builder) WithShootObject(shootObject *gardencorev1beta1.Shoot) *Builder {
+	b.shootObjectFunc = func() (*gardencorev1beta1.Shoot, error) { return shootObject, nil }
+	return b
+}
+
+// WithShootObjectFromLister sets the shootObject attribute at the Builder after fetching it from the given lister.
+func (b *Builder) WithShootObjectFromLister(shootLister gardencorelisters.ShootLister, namespace, name string) *Builder {
+	b.shootObjectFunc = func() (*gardencorev1beta1.Shoot, error) { return shootLister.Shoots(namespace).Get(name) }
+	return b
+}
+
+// WithCloudProfileObject sets the cloudProfileFunc attribute at the Builder.
+func (b *Builder) WithCloudProfileObject(cloudProfileObject *gardencorev1beta1.CloudProfile) *Builder {
+	b.cloudProfileFunc = func(string) (*gardencorev1beta1.CloudProfile, error) { return cloudProfileObject, nil }
+	return b
+}
+
+// WithCloudProfileObjectFromLister sets the cloudProfileObject attribute at the Builder after fetching it from the given lister.
+func (b *Builder) WithCloudProfileObjectFromLister(cloudProfileLister gardencorelisters.CloudProfileLister) *Builder {
```

Comment:
Can we remove this method in favor of adding a default implementation to `Build()`, that gets the cloudprofile, if no `cloudProfileFunc` has been set, or something similar?

Reply:
Why?

Reply:
FMPOV, calling this method and passing only an extra lister is unnecessary, as you have to pass a client to `Build` anyways. If you call this method, you're basically expressing, that you don't have the `CloudProfile` object yet and want it to be retrieved on `Build()`. This could be done implicitly (by not calling `WithCloudProfileObject`) to remove the duplication of clients.

Reply:
See https://github.com/gardener/gardener/pull/2217#issuecomment-620994942 for a comment about default implementations

Reply:
Yes, but the shoot builder's `Build` method anyways takes a client.

Reply:
Also obsolete with https://github.com/gardener/gardener/pull/2217#discussion_r417100845, as we rather want to use the lister's cache instead of making unnecessary external API calls.